auth/password_hash_test.go

```diff
@@ -64,3 +67,394 @@ func TestSetBcryptCost(t *testing.T) {
+	const maxCacheSize = 5
+	var tests = []struct {
+		name      string
+		cache     Cache
+		cacheType cacheType
+	}{
+		{"RandReplKeyCache", NewRandReplKeyCache(maxCacheSize), randReplKeyCache},
+		{"NoReplKeyCache", NewNoReplKeyCache(maxCacheSize), noReplKeyCache},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			var wg sync.WaitGroup
+			cache := tt.cache
+			defer func() {
+				cache.Purge()
+				assert.Equal(t, 0, cache.Len())
+			}()
+
+			for i := 0; i < 10; i++ {
+				wg.Add(1)
+				go func(cache Cache, i int) {
+					defer wg.Done()
+					key := fmt.Sprintf("k%d", i)
+					cache.Put(key)
+					base.Infof(base.KeyAuth, "Goroutine%v Put key: %v", i, key)
+				}(cache, i)
+
+				wg.Add(1)
+				go func(cache Cache, i int) {
+					defer wg.Done()
+					key := fmt.Sprintf("k%d", i)
+					ok := cache.Contains(key)
+					base.Infof(base.KeyAuth, "Goroutine%v Contains: %v, key: %v", i, ok, key)
+				}(cache, i)
+			}
+			wg.Wait()
+			assert.Equal(t, 5, cache.Len())
+		})
+	}
+}
+
+func BenchmarkPutAndOverflow(b *testing.B) {
+	const maxCacheSize = 100
+
+	setupCache := func(cacheType cacheType, warmupCache bool) (cache Cache) {
+		if cacheType == randReplKeyCache {
+			cache = NewRandReplKeyCache(maxCacheSize)
+		} else if cacheType == noReplKeyCache {
+			cache = NewNoReplKeyCache(maxCacheSize)
+		}
+		if warmupCache {
+			for i := 0; i < maxCacheSize; i++ {
+				cache.Put(key())
+			}
+		}
+		return cache
+	}
+
+	benchmarks := []struct {
+		name        string
+		cacheType   cacheType
+		warmupCache bool
+		runParallel bool
+	}{
+		{"FillRandReplKeyCache", randReplKeyCache, false, false},
+		{"FillNoReplKeyCache", noReplKeyCache, false, false},
+
+		{"RefillRandReplKeyCache", randReplKeyCache, true, false},
+		{"RefillNoReplKeyCache", noReplKeyCache, true, false},
+
+		{"ParallelFillRandReplKeyCache", randReplKeyCache, false, true},
+		{"ParallelFillNoReplKeyCache", noReplKeyCache, false, true},
+
+		{"ParallelRefillRandReplKeyCache", randReplKeyCache, true, true},
+		{"ParallelRefillNoReplKeyCache", noReplKeyCache, true, true},
+	}
+	for _, bm := range benchmarks {
+		b.Run(bm.name, func(b *testing.B) {
+			cache := setupCache(bm.cacheType, bm.warmupCache)
+			b.ResetTimer()
+			if bm.runParallel {
+				b.RunParallel(func(pb *testing.PB) {
+					b.ReportAllocs()
+					for pb.Next() {
+						cache.Put(key())
+					}
+				})
+			} else {
+				b.Run(bm.name, func(b *testing.B) {
+					b.ReportAllocs()
+					for i := 0; i < b.N; i++ {
+						cache.Put(key())
+					}
+				})
+			}
+		})
+	}
+}
+
+func BenchmarkContains(b *testing.B) {
+	const maxCacheSize = 1
+	benchmarks := []struct {
+		name  string
+		cache Cache
+	}{
+		{"RandReplKeyCache", NewRandReplKeyCache(maxCacheSize)},
+		{"NoReplKeyCache", NewNoReplKeyCache(maxCacheSize)},
+	}
+	for _, bm := range benchmarks {
+		b.Run(bm.name, func(b *testing.B) {
+			b.ReportAllocs()
+			cache := bm.cache
+			cache.Put("foo")
```

Comment:
This benchmark would be more effective if it was using a populated cache.

Reply:
This benchmark has been refactored with populated cache.